proxy/http/remap/NextHopSelectionStrategy.cc

```diff
@@ -0,0 +1,412 @@
+constexpr std::string_view hash_key_path_query    = "path+query";
+constexpr std::string_view hash_key_path_fragment = "path+fragment";
+constexpr std::string_view hash_key_cache         = "cache_key";
+
+// ring mode strings
+constexpr std::string_view alternate_rings = "alternate_ring";
+constexpr std::string_view exhaust_rings   = "exhaust_ring";
+
+// health check strings
+constexpr std::string_view active_health_check  = "active";
+constexpr std::string_view passive_health_check = "passive";
+
+constexpr const char *policy_strings[] = {"NH_UNDEFINED", "NH_FIRST_LIVE", "NH_RR_STRICT",
+                                          "NH_RR_IP",     "NH_RR_LATCHED", "NH_CONSISTENT_HASH"};
+
+NextHopSelectionStrategy::NextHopSelectionStrategy(const std::string_view &name, const NHPolicyType &policy)
+{
+  strategy_name = name;
+  hash_key      = NH_PATH_HASH_KEY;
+  policy_type   = policy;
+  scheme        = NH_SCHEME_NONE;
+  ring_mode     = NH_ALTERNATE_RING;
+  groups        = 0;
+  num_parents   = 0;
+  NH_Debug(NH_DEBUG_TAG, "Using a selection strategy of type %s", policy_strings[policy]);
+}
+
+//
+// parse out the data for this strategy.
+//
+bool
+NextHopSelectionStrategy::Init(const YAML::Node &n)
+{
+  NH_Debug(NH_DEBUG_TAG, "calling Init()");
+  try {
+    if (n["hash_key"]) {
+      auto hash_key_val = n["hash_key"].Scalar();
+      if (hash_key_val == hash_key_uri) {
+        hash_key = NH_URI_HASH_KEY;
+      } else if (hash_key_val == hash_key_url) {
+        hash_key = NH_URL_HASH_KEY;
+      } else if (hash_key_val == hash_key_hostname) {
+        hash_key = NH_HOSTNAME_HASH_KEY;
+      } else if (hash_key_val == hash_key_path) {
+        hash_key = NH_PATH_HASH_KEY;
+      } else if (hash_key_val == hash_key_path_query) {
+        hash_key = NH_PATH_QUERY_HASH_KEY;
+      } else if (hash_key_val == hash_key_path_fragment) {
+        hash_key = NH_PATH_FRAGMENT_HASH_KEY;
+      } else if (hash_key_val == hash_key_cache) {
+        hash_key = NH_CACHE_HASH_KEY;
+      } else {
+        hash_key = NH_PATH_HASH_KEY;
+        NH_Note("Invalid 'hash_key' value, '%s', for the strategy named '%s', using default '%s'.", hash_key_val.c_str(),
+                strategy_name.c_str(), hash_key_path.data());
+      }
+    }
+
+    if (n["scheme"]) {
+      auto scheme_val = n["scheme"].Scalar();
+      if (scheme_val == "http") {
+        scheme = NH_SCHEME_HTTP;
+      } else if (scheme_val == "https") {
+        scheme = NH_SCHEME_HTTPS;
+      } else {
+        scheme = NH_SCHEME_NONE;
+        NH_Note("Invalid 'scheme' value, '%s', for the strategy named '%s', setting to NH_SCHEME_NONE", scheme_val.c_str(),
+                strategy_name.c_str());
+      }
+    }
+
+    // go_direct config.
+    if (n["go_direct"]) {
+      go_direct = n["go_direct"].as<bool>();
+    }
+
+    // parent_is_proxy config.
+    if (n["parent_is_proxy"]) {
+      parent_is_proxy = n["parent_is_proxy"].as<bool>();
```

Comment:
this value is never used. What is its intended purpose?

Reply:
@a-canary it's used in HttpTransact::parent_is_proxy().  It comes from parent selection and indicates that the parent is a trafficserver cache or not.  the nexthop could be an origin.  Mainly it's used to strip the fqdn from the request when the nexthop is an origin.